File: reference/dotnet/Adapters/Company.Product.Adapters.Rest.Generated/Controllers/AbstractAnimalsApi.cs
namespace Company.Product.Adapters.Rest.Controllers;

public abstract class AbstractAnimalsApi : ControllerBase
{
    private readonly ICreateAnimalUseCase createAnimalUseCase;
    private readonly IGetAnimalUseCase getAnimalUseCase;
    private readonly IGetAnimalsUseCase getAnimalsUseCase;

    protected AbstractAnimalsApi(ICreateAnimalUseCase createAnimalUseCase, IGetAnimalUseCase getAnimalUseCase, IGetAnimalsUseCase getAnimalsUseCase)
    {
        this.createAnimalUseCase = createAnimalUseCase ?? throw new ArgumentNullException(nameof(createAnimalUseCase));
        this.getAnimalUseCase = getAnimalUseCase ?? throw new ArgumentNullException(nameof(getAnimalUseCase));
        this.getAnimalsUseCase = getAnimalsUseCase ?? throw new ArgumentNullException(nameof(getAnimalsUseCase));
    }

    [HttpPost]
    [Route("/animals")]
    public async virtual Task<ActionResult> CreateAnimal([FromBody][Required] Animal animal, CancellationToken cancellationToken)
    {
        await createAnimalUseCase.CreateAnimal(animal: animal.ToDomain(), cancellationToken: cancellationToken);
        return Empty;
    }

    [HttpGet]
    [Route("/animals")]
    public async virtual Task<ActionResult<GetAnimalsResponse>> GetAnimals([FromQuery] int limit, [FromQuery] int offset, CancellationToken cancellationToken)
    {
        var animals = await getAnimalsUseCase.GetAnimals(limit: limit, offset: offset, cancellationToken: cancellationToken);

        return new GetAnimalsResponse()
        {
            Animals = animals.Select(a => AnimalMapper.FromDomain(a))
        };
    }

    [HttpGet]
    [Route("/animals/{animalId}")]
    public async virtual Task<ActionResult<GetAnimalResponse>> GetAnimal(Guid animalId, CancellationToken cancellationToken)
    {
        var animal = await getAnimalUseCase.GetAnimal(animalId: animalId, cancellationToken: cancellationToken);

        return new GetAnimalResponse()
        {
            Animal = AnimalMapper.FromDomain(animal)
        };
    }
}

File: reference/springboot/adapters-rest-generated/src/main/java/org/example/adapters/rest/AbstractAnimalsApi.java
package org.example.adapters.rest;

import lombok.RequiredArgsConstructor;

import org.example.adapters.rest.dto.AnimalDto;
import org.example.adapters.rest.dto.GetAnimalResponse;
import org.example.adapters.rest.dto.GetAnimalsResponse;
import org.example.domain.usecases.CreateAnimalUseCase;
import org.example.domain.usecases.GetAnimalUseCase;
import org.example.domain.usecases.GetAnimalsUseCase;
import org.springframework.http.ResponseEntity;
import java.util.List;
import java.util.UUID;

@RequiredArgsConstructor
public abstract class AbstractAnimalsApi implements AnimalsApi {
    private final GetAnimalUseCase getAnimalUseCase;
    private final GetAnimalsUseCase getAnimalsUseCase;
    private final CreateAnimalUseCase createAnimalUseCase;

    @Override
    public ResponseEntity<Void> createAnimal(AnimalDto animal) {
        createAnimalUseCase.createAnimal(animal.toDomain());
        return map(animal);
    }

    @Override
    public ResponseEntity<GetAnimalResponse> getAnimal(UUID animalId) {
        var animal = getAnimalUseCase.getAnimal(animalId);
        return map(animalId, AnimalDto.fromDomain(animal));
    }

    @Override
    public ResponseEntity<GetAnimalsResponse> getAnimals(int limit, int offset) {
        //discrepancy, map explicit limit/offset to best-practice Page/Pageable
        var animals = getAnimalsUseCase.getAnimals(limit, offset);

       return map(
            limit, 
            offset, 
            animals.stream()
                .map(AnimalDto::fromDomain)
                .toList()
        );
    }

    protected abstract ResponseEntity<Void> map(AnimalDto animal);
    protected abstract ResponseEntity<GetAnimalResponse> map(UUID animalId, AnimalDto useCaseResponse);
    protected abstract ResponseEntity<GetAnimalsResponse> map(int limit, int offset, List<AnimalDto> useCaseResponse);
}
